voip_utils/sip.py

```diff
@@ -190,87 +270,203 @@ def datagram_received(self, data: bytes, addr):
+                caller_rtp_port: Optional[int] = None
+                opus_payload_type = OPUS_PAYLOAD_TYPE
+                body_lines = body.splitlines()
+                for line in body_lines:
+                    line = line.strip()
+                    if line:
+                        key, value = line.split("=", maxsplit=1)
+                        if key == "m":
+                            parts = value.split()
+                            if parts[0] == "audio":
+                                caller_rtp_port = int(parts[1])
+                        elif key == "a" and value.startswith("rtpmap:"):
+                            # a=rtpmap:123 opus/48000/2
+                            codec_str = value.split(":", maxsplit=1)[1]
+                            codec_parts = codec_str.split()
+                            if (len(codec_parts) > 1) and (
+                                codec_parts[1].lower().startswith("opus")
+                            ):
+                                opus_payload_type = int(codec_parts[0])
+                                _LOGGER.debug(
+                                    "Detected OPUS payload type as %s",
+                                    opus_payload_type,
+                                )
+
+                if caller_rtp_port is None:
+                    raise VoipError("No caller RTP port")
+
+                # Extract host from ruri
+                # sip:user@123.123.123.123:1234
+                re_splituri = re.compile(
+                    r"(?P<scheme>\w+):"  # Scheme
+                    + r"(?:(?P<user>[\w\.]+):?(?P<password>[\w\.]+)?@)?"  # User:Password
+                    + r"\[?(?P<host>"  # Begin group host
+                    + r"(?:\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3})|"  # IPv4 address Host Or
+                    + r"(?:(?:[0-9a-fA-F]{1,4}):){7}[0-9a-fA-F]{1,4}|"  # IPv6 address Host Or
+                    + r"(?:(?:[0-9A-Za-z]+\.)+[0-9A-Za-z]+)"  # Hostname string
+                    + r")\]?:?"  # End group host
+                    + r"(?P<port>\d{1,6})?"  # port
+                    + r"(?:\;(?P<params>[^\?]*))?"  # parameters
+                    + r"(?:\?(?P<headers>.*))?"  # headers
                 )
-            )
+                re_uri = re_splituri.search(ruri)
+                if re_uri is None:
+                    raise ValueError("Receiver URI did not match expected pattern")
+
+                server_ip = re_uri.group("host")
+                if not is_ipv4_address(server_ip):
+                    raise VoipError(f"Invalid IPv4 address in {ruri}")
+
+                self.on_call(
+                    CallInfo(
+                        caller_endpoint=caller_endpoint,
+                        caller_rtp_port=caller_rtp_port,
+                        server_ip=server_ip,
+                        headers=headers,
+                        opus_payload_type=opus_payload_type,
+                    )
+                )
+            elif method == "sip/2.0":
```

Comment:
is this really the method, or should `parse_sip` differentiate between protocol messages?